### DISCUSSION TITLE
Aws::ivsrealtime::Model::CreateParticipantTokenRequest Appears to Corrupt Memory?

### DISCUSSION THREAD

**Discussion by AngusMac91** [aws/aws-sdk-cpp#3137] [Q&A] [answered]
Hi all,

I'm trying to create a basic C++ app that will grab a new participant token from a stage for IVS Realtime. However, I'm having some assertion issues when using the CreateParticipantTokenRequest object, and also the CreateParticipantTokenOutcome object (which is the response returned). Both appear to corrupt memory when deleted, even when they haven't been used and its the only piece of code in place.
e.g.
```
Aws::SDKOptions options;
options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Trace;
Aws::InitAPI(options);
{
	{
		Aws::ivsrealtime::Model::CreateParticipantTokenRequest request;
	} //Crashes here
	....
}
Aws::ShutdownAPI(options);
```
![image](https://github.com/user-attachments/assets/0b29ea23-954d-4d90-8395-f7d2febbeb82)

Has anyone noticed this, or is there potentially something wrong with my build of the SDK?

**Comment by sbiscigl** [comment]
> or is there potentially something wrong with my build of the SDK?

I think there is something wrong with your environment or build of the SDK. Anecdotally i was not able to reproduce the issue you had using a MSVC toolchain, but also I was not able to reproduce your issue when running from a al-2023 docker image.

Project Structure:
```
~/sdk-usage-workspace ❯❯❯ tree -L 1
.
├── CMakeLists.txt
├── Dockerfile
├── main.cpp
└── replicate.sh
```

Dockerfile
```Dockerfile
# Using offical Amazon Linux 2023 image from public ECR
FROM public.ecr.aws/amazonlinux/amazonlinux:2023

# Install compiler et al.
RUN yum groupinstall "Development Tools" -y

# Install required dependencies
RUN yum install -y ninja-build cmake3 curl-devel openssl-devel

# Install sdk
RUN git clone --depth 1 --recurse-submodules https://github.com/aws/aws-sdk-cpp && \
    cd aws-sdk-cpp && \
    mkdir build && \
    cd build && \
    cmake -G Ninja -DBUILD_ONLY="ivs-realtime" \
      -DAUTORUN_UNIT_TESTS=OFF .. && \
    cmake --build . && \
    cmake --install .

## Copy over and build
RUN mkdir sdk-example
COPY CMakeLists.txt /sdk-example/CMakeLists.txt
COPY main.cpp /sdk-example/main.cpp
RUN cd sdk-example &&\
    mkdir build &&\
    cd build &&\
    cmake -G Ninja .. && \
    cmake --build .

```

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.13)
project(sdk_usage_workspace)
set(CMAKE_CXX_STANDARD 20)
find_package(AWSSDK REQUIRED COMPONENTS ivs-realtime)
add_executable(${PROJECT_NAME} "main.cpp")
target_link_libraries(${PROJECT_NAME} PRIVATE ${AWSSDK_LINK_LIBRARIES})
```

main.cpp
```cpp
#include <aws/core/Aws.h>
#include <aws/ivs-realtime/model/CreateParticipantTokenRequest.h>

auto main() -> int
{
    Aws::SDKOptions options;
    options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Trace;
    Aws::InitAPI(options);
    {
        {
            Aws::ivsrealtime::Model::CreateParticipantTokenRequest request;
        }
    }
    Aws::ShutdownAPI(options);
    return 0;
}

```

replicate.sh
```shell
#!/bin/zsh

set -u

# build image
docker build -t test-image .

# run example
docker run --name test-image test-image /sdk-example/build/sdk_usage_workspace
```

when we run with `./replicate.sh` while it does seem to be working as expected, we would be more than happy to help you figure out whats wrong in your environment if you can give us reproducible steps to create it, preferably a dockerfile to remove any "on my machine" issues.

**Reply by AngusMac91** [reply]
Hey, 

Thanks for getting back to me on this. I'm not totally savvy with docker, but I can give exact steps as to how I've created the SDK on my current machine with logs produced by CMake etc. As far as I can see the build succeeds without any glaring issues, but maybe I'm missing something obvious?

So via cmnd prompt on Windows I'm effectively following the steps from this page https://sdk.amazonaws.com/cpp/api/LATEST/root/html/index.html:
```
cd D:/AWS_SDK/source
git clone --recurse-submodules https://github.com/aws/aws-sdk-cpp
cd ../build
cmake ..\source\aws-sdk-cpp -DCMAKE_BUILD_TYPE=Debug -DBUILD_ONLY="ivs-realtime" -DCMAKE_INSTALL_PREFIX="../install" -DCPP_STANDARD=14
cmake --build . --config=Debug
cmake --install . --config=Debug
```
This leaves me with an install folder that contains lib, bin, and include.

I'm then placing these in a directory next to my Visual Studio solution and changing the solutions properties:
Properties -> C/C++ -> General -> Additional Include Directories - Added path to AWS SDK include folder
Properties -> C/C++ -> Preprocessor -> Preprocessor Definitions - Added USE_IMPORT_EXPORT (as stated by this page: https://sdk.amazonaws.com/cpp/api/LATEST/root/html/md_docs_2_s_d_k__usage__guide.html)
Properties -> Linker -> General -> Additional Library Directories - Added path to AWS SDK lib folder
Properties -> Linker -> Input -> Additional Dependencies - Added the two required lib files, aws-cpp-sdk-core.lib;aws-cpp-sdk-ivs-realtime.lib;

This then allows me to include, compile, and run my project with AWS code; however with the issue stated in the original post. I've attached some of the CMake files I believe to be relevant but let me know if I've missed anything.
[CMakeConfigureLog.txt](https://github.com/user-attachments/files/17307079/CMakeConfigureLog.txt)
[CMakeCache.txt](https://github.com/user-attachments/files/17307090/CMakeCache.txt)
[CMake_Log.txt](https://github.com/user-attachments/files/17307108/CMake_Log.txt)

**Reply by AngusMac91** [reply]
Sorry I should probably clarify - I'm trying to link the AWS sdk into a DLL project, not a base C++ executable.

**Reply by jmklix** [reply]
Can you make sure you have the same Microsoft C runtime in both environments

**Reply by AngusMac91** [reply, marked as answer]
Hey, sorry I should have updated this discussion when I solved this a couple of weeks ago. Yes this was caused by a mismatch between C runtime environments. The SDK was being built dynamically linking to CRT, but my software was statically linking to it. Aligning those has resolved the issue.
